Editorial review: Update information on the Webauthn hints option

chrisdavidmills opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. 🙌

Add details below to help us review your pull request (PR).
Explain your changes and link to a related issue or pull request.
Your PR may be delayed or closed if you don't provide enough information. -->

### Description

Chrome 128 adds support for a new `hints` option that can be specified inside the `publicKey` option of a Webauthn `create()` or `get()` call, to hint to the browser what UI to show to the user for creating or authenticating with a security key. For example, do their users tend to use a physical security key like a YubiKey, or a smartphone-based authenticator app?

The `hints` option was already covered on both the `PublicKeyCredentialCreationOptions` and `PublicKeyCredentialRequestOptions` dictionary reference pages. This PR expands and improves on the provided information.

<!-- ✍️ Summarize your changes in one or two sentences. -->

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details

<!-- 🔗 Link to release notes, browser docs, bug trackers, source control, or other resources. -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request must be merged first, use "**Depends on:** #123" -->

<!-- 🔎 After submitting, the 'Checks' tab of your PR shows the build status. -->

chrisdavidmills committed:
Update information on the Webauthn hints option

chrisdavidmills committed:
Merge branch 'main' into webauthn-hints

nsatragno commented on files/en-us/web/api/publickeycredentialcreationoptions/index.md:
```diff
@@ -95,6 +95,23 @@ The **`PublicKeyCredentialCreationOptions`** dictionary represents the object pa
+
+    When multiple strings are included in the array, their order denotes the order of preference, from high to low. Supporting browsers that respect the hints should use the first one that they understand.
+
+    The `hints` option provides a more flexible way to specify UI preferences for creating a credential than the [`authenticatorSelection.authenticatorAttachment`](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#authenticatorattachment) option, which completely hides the non-chosen option. Specified `hints` may contradict hints provided in both the `authenticatorAttachment` and [`transport ...
```
> and [`transports`](#transports) options

This is only really relevant for get assertion. For make credential, `transports` are only present for the exclude credentials list (which represents devices the user already registered, so the RP intention is to not override those credentials), so there's no "contradiction". I'd remove this bit.

chrisdavidmills replied:
OK, understood. I've updated this to:

> Specified `hints` may contradict hints provided in the `authenticatorAttachment` option. When the provided `hints` contradict this option, the `hints` take precedence.

nsatragno commented on files/en-us/web/api/publickeycredentialcreationoptions/index.md:
```diff
@@ -95,6 +95,23 @@ The **`PublicKeyCredentialCreationOptions`** dictionary represents the object pa
+    - `"client-device"`
+      - : The UI should recommend using an authenticator available on the same device they are using to access the RP client to create the credential. It is analogous to the `authenticatorAttachment` [`platform`](#platform) value.
+    - `"hybrid"`
+      - : The UI should recommend using a general-purpose authenticator, such as a smartphone-based authenticator app, to create the credential. This favors using a cross-device approach to handling authentication, relying on a combination of laptop and smartphone, for example. It is analogous to the `authenticatorAttachment` [`cross-platform`](#cross-platform) value.
```
> It is analogous to the `authenticatorAttachment` [`cross-platform`](#cross-platform) value.

This is also not quite accurate. `attachment: "cross-platform"` is essentially a combination of `security-key` and `hybrid`. E.g. if a device doesn't have bluetooth and an RP specifies `attachment: "cross-platform"`, the resulting UI is very similar to `hints: "security-key"`.

chrisdavidmills replied:
Cool, makes sense. I've updated the text to remove the bit you highlighted above, and include a variant of your text in the notes below:

> The `authenticatorAttachment` [`cross-platform`](#cross-platform) value is essentially a combination of the `hints` option `security-key` and `hybrid` values — if a device doesn't have bluetooth and an RP specifies `attachment: "cross-platform"`, the resulting UI will likely be similar to the `hints: "security-key"` UI.

I've said "will likely be similar", making it a bit less definite, as theoretically other implementations may come along in the future that have a different UI from the Chrome implementation.

nsatragno commented on files/en-us/web/api/publickeycredentialrequestoptions/index.md:
```diff
@@ -41,15 +42,21 @@ It is used to request a {{domxref("PublicKeyCredential")}} provided by an {{glos
+
+    When multiple strings are included in the array, their order denotes the order of preference, from high to low. Supporting browsers that respect the hints should use the first one that they understand.
+
+    The `hints` option may contradict hints provided in the [`transports`](#transports) option. When the provided `hints` contradict this option, the `hints` take precedence.
```
If you accept the suggestion above, please also update this section accordingly.

nsatragno reviewed (commented):
Thanks for your contribution! Left some minor comments.

chrisdavidmills commented on files/en-us/web/api/publickeycredentialcreationoptions/index.md:
```diff
@@ -95,6 +95,23 @@ The **`PublicKeyCredentialCreationOptions`** dictionary represents the object pa
+
+    When multiple strings are included in the array, their order denotes the order of preference, from high to low. Supporting browsers that respect the hints should use the first one that they understand.
+
+    The `hints` option provides a more flexible way to specify UI preferences for creating a credential than the [`authenticatorSelection.authenticatorAttachment`](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#authenticatorattachment) option, which completely hides the non-chosen option. Specified `hints` may contradict hints provided in both the `authenticatorAttachment` and [`transport ...
```
These sound good. I've included them pretty much verbatim, but I've split the paragraph in half so it is easier to read.

chrisdavidmills committed:
Fixes for nsatragno review comments

chrisdavidmills commented:
Thanks so much for the tech review, @nsatragno! I have addressed your comments. Please let me know if you think it needs any more changes, or give me an LGTM if you are satisfied.

nsatragno reviewed (commented):
✅ LGTM, thanks!

chrisdavidmills commented:
@hamishwillee, are you OK to give this one an editorial review? You've got previous WebAuthn experience, and it is only a smallish one.

hamishwillee reviewed (approved):
@chrisdavidmills This seems very clear and concise to me. Thank you.

FWIW not a blocker for this, but it would be great if we had a "Using" guide for this that covered things like hints (the kind of information we see in https://developer.chrome.com/blog/passkeys-updates-chrome-129#hints)

chrisdavidmills commented:
> @chrisdavidmills This seems very clear and concise to me. Thank you.

Nice one!

> FWIW not a blocker for this, but it would be great if we had a "Using" guide for this that covered things like hints (the kind of information we see in https://developer.chrome.com/blog/passkeys-updates-chrome-129#hints)

Yeah, I agree that we need better guides to using this stuff. The holistic usage picture of WebAuthn is still not clear.
